crates/ibc/src/clients/ics07_tendermint/client_state.rs

```diff
@@ -876,13 +854,161 @@ impl Ics2ClientState for ClientState {
+
+        // Note: verification of proofs that unmarshalled correctly has been done
+        // while decoding the proto message into a `MsgEnvelope` domain type
+        let merkle_proof_upgrade_client = MerkleProof::from(proof_upgrade_client);
+        let merkle_proof_upgrade_cons_state = MerkleProof::from(proof_upgrade_consensus_state);
+
+        // Make sure the latest height of the current client is not greater then
+        // the upgrade height This condition checks both the revision number and
+        // the height
+        if self.latest_height() >= upgraded_tm_client_state.latest_height() {
+            return Err(ClientError::LowUpgradeHeight {
+                upgraded_height: self.latest_height(),
+                client_height: upgraded_tm_client_state.latest_height(),
+            });
+        }
+
+        // Ensure that the new unbonding period is not shorter than the current's
+        // client unbonding period
+        if self.unbonding_period > upgraded_tm_client_state.unbonding_period {
+            return Err(ClientError::ClientSpecific {
+                description:
+                    "cannot upgrade client with a shorter unbonding period than the current one"
+                        .to_string(),
+            });
+        }
+
+        // Check to see if the upgrade path is set
+        let mut upgrade_path = upgraded_tm_client_state.clone().upgrade_path;
+        if upgrade_path.pop().is_none() {
+            return Err(ClientError::ClientSpecific {
+                description: "cannot upgrade client as no upgrade path has been set".to_string(),
+            });
+        };
+
+        let last_height = self.latest_height().revision_height();
+
+        // Construct the merkle path for the client state
+        let mut client_upgrade_path = upgrade_path.clone();
+        client_upgrade_path.push(ClientUpgradePath::UpgradedClientState(last_height).to_string());
+
+        let client_upgrade_merkle_path = MerklePath {
+            key_path: client_upgrade_path,
+        };
+
+        upgraded_tm_client_state.zero_custom_fields();
+        let client_state_value =
+            Protobuf::<RawTmClientState>::encode_vec(&upgraded_tm_client_state)
+                .map_err(ClientError::Encode)?;
+
+        // Verify the proof of the upgraded client state
+        merkle_proof_upgrade_client
+            .verify_membership(
+                &self.proof_specs,
+                root.clone().into(),
+                client_upgrade_merkle_path,
+                client_state_value,
+                0,
+            )
+            .map_err(ClientError::Ics23Verification)?;
+
+        // Construct the merkle path for the consensus state
+        let mut cons_upgrade_path = upgrade_path;
+        cons_upgrade_path
+            .push(ClientUpgradePath::UpgradedClientConsensusState(last_height).to_string());
+        let cons_upgrade_merkle_path = MerklePath {
+            key_path: cons_upgrade_path,
+        };
```

Comment:
Similarly here, this doesn't seem to be [the same as ibc-go](https://github.com/cosmos/ibc-go/blob/c0acd5bd1778f2b7ecdf593006f56bd3e273bd49/modules/light-clients/07-tendermint/types/upgrade.go#L144-L156).

Reply:
First, I coded this the same as ibc-go, but then found out on our side we access something simpler. So, one difference here is that I tried to use the internal structs to `ibc-rs`, and created the path with `ClientUpgradePath::UpgradedClientConsensusState(last_height)`. 
I didn't find such an instance on `ibc-go`
But truly it should be verified with tests

Reply:
> So, one difference here is that I tried to use the internal structs to ibc-rs, and created the path with ClientUpgradePath::UpgradedClientConsensusState(last_height)

This part of ibc-rs hasn't been touched in a while and it is likely outdated and/or incorrect (like a lot of things were just a few months ago). If an internal struct is incorrect or outdated, we should just change it.

Reply:
#391